apache/superset
# feat: Improves SafeMarkdown HTML sanitization

### SUMMARY
Currently, we are rendering HTML by default and using the following function to sanitize the content:

```
function isSafeMarkup(node: MarkdownAbstractSyntaxTree) {
  return node.type === 'html' && node.value
    ? !/(href|src)="(javascript|vbscript|file):.*"/gim.test(node.value)
    : true;
}
```

This only blocks a small subset of potential attack vectors as you can easily see [here](https://cheatsheetseries.owasp.org/cheatsheets/XSS_Filter_Evasion_Cheat_Sheet.html). To improve the component's security this PR does the following:

- Upgrades [react-markdown](https://github.com/remarkjs/react-markdown) from version 4.3.1 to 8.0.3. This will help with many security issues and also enable the use of some really cool plugins to extend our markdown capabilities.
- Installs [rehype-raw](https://github.com/rehypejs/rehype-raw) to parse HTML content.
- Installs [rehype-sanitize](https://github.com/rehypejs/rehype-sanitize) to sanitize the HTML content. By default, it uses the same schema used by GitHub to sanitize HTML, which means a more secure configuration.
- Adds the `HTML_SANITIZATION` configuration option to enable/disable HTML sanitization.
- Adds the `HTML_SANITIZATION_SCHEMA_EXTENSIONS` configuration option to extend the default schema in case someone has a controlled environment and wishes to use some blocked elements/attributes.

### TESTING INSTRUCTIONS
- Make sure HTML is sanitized or not depending on the value of `HTML_SANITIZATION`
- Make sure HTML sanitization reflects the extensions made in `HTML_SANITIZATION_SCHEMA_EXTENSIONS`
- Make sure the examples that use markdown are correctly rendered

### ADDITIONAL INFORMATION
